### DISCUSSION TITLE
Convert v1 torrent to hybrid torrent

### DISCUSSION THREAD

**Discussion by cherinyy** [transmission/transmission#4148] [Ideas]
Hello,

_I am posting my idea here because of an error in the forum program that prevented me from registering for a forum account._

I would like to know if it is possible to convert v1 torrent to hybrid torrent?

I think the best thing about BitTorrent v2 is that it assigns an identifier to each file, now there are too many files being repackaged over and over again.

If the v1 torrent can be converted to a hybrid torrent, this allows older files to benefit as well, and some torrents that were thought to be dead will be revived.

**Comment by davidnicol** [comment]
I was actually stunned that the v2 spec didn't include a migration path kind of thing. It seems to me that the conversion should be transparent and automatic, by crafting v2 hashes for all the files in anything the client is seeding and announcing them to MainNet every 200-300 minutes. There is something of an executive focus problem though -- which comes first, adding the feature to Transmission or adding v2fs to IPFS? Or folding the laundry  and mowing the yard? There may need to be another BEP for messages stating what the v2 hash is for a sought  v1 hash, like peer exchange. I think that the ideal roll-out of v2 would be an acceleration system transparently deduplicating files from v1 torrents, without  having to craft hybrid magnet links. As it stands, v2 is a fine example of "mousetrap syndrome." It's a technically superior solution with very little adoption.

**Comment by davidnicol** [comment]
reviewing the relevant BEPs, I imagine a new extension ??_v2migrate which, when supported on both sides, defines a message which would be essentially the BEP 52 file tree for all of the parts of the v1 torrent that a sharing peer or seeder has.

The v2_migrate extension, seeing the v2 format as an accelerator for v1 torrents, will need to define alternatives to some of the peer wire protocol as described in BEP 52 -- in particular, the piece and request messages are ambiguous unless a v1 torrent is padded because the versions are taking hashes of different things to describe the pieces in the info dictionary.  Therefore the v2migrate extension provides an altenative to

> Implementations supporting both formats can join both swarms by calculating the new and old infohashes and downloading them to the same storage. Before doing so they must validate that the content (file names, order, piece alignment) is identical. During the download they must also verify that pieces match both piece hash formats. If any inconsistency is detected they may either abort or fall back to downloading one of the two formats as if the other were not present.
> 
> When initiating a connection and sending the sha1 infohash of such a hybrid torrent a peer can set the 4th most significant bit in the last byte of the reserved bitfield to indicate that it also supports the new format. The remote peer may then respond with the new infohash to upgrade the connect to the new format.

Instead, v2migrate defines facility for v2-style transfer at file grain within the context of a v1 torrent. So there is no defined "piece size" which is an abstraction to make large transfers have a reasonable number of pieces. Instead of clients being only required to 

> support setting the hash request base layer to the leaf and piece layers,

 for BEP52 conformance, a v2migrate conformant peer MUST honor arbitrary base layers in hash request requests and respond as requested in hashes responses. Except that the peer MUST issue hash-reject messages in response to hash-request messages for data a peer does not have to share. A peer which is currently gathering data based on alleged hashes SHOULD send a BEP6  suggest-piece message or something similar to it describing leaves the peer has verified against the v1 manifest.

Instead of providing new v2 request and piece message types, the semantics of request, piece, and bitfield switch to the index number referring to an index into the the most recently received hashes list. Instead of only being sent at the beginning of a session, bitfield may be sent following a hashes message with base layer 0 to effectively request all leaves corresponding to zero bits.

The v2migrate extension therefore needs to define a new message to reset the session semantics back to v1/BEP3 semantics on completion of downloading v2 leaf pieces from individual files referred to by their pieces_root hashes without regard to how they might appear in an infohash describing their curation.

In conclusion, that was an interesting three hour message composition while re-reading [The BitTorrent Protocol Specification v2, last modified August 31, 2017](http://www.bittorrent.org/beps/bep_0052.html)

**Reply by cherinyy** [reply]
This is a great idea! Would love for you to post your ideas on the mailing list: https://www.bittorrent.org/mailing_list.html

Or on [bittorrent/bittorrent.org/issues](https://github.com/bittorrent/bittorrent.org/issues)
